src/lem/constrainer.rs

```diff
@@ -44,14 +44,70 @@ impl<F: LurkField> AllocationManager<F> {
             }
         }
     }
+
+    /// Calls `get_or_alloc_num` to allocate tag and hash for a pointer.
+    pub(crate) fn get_or_alloc_ptr<CS: ConstraintSystem<F>>(
+        &mut self,
+        cs: &mut CS,
+        z_ptr: &ZPtr<F>,
+    ) -> Result<AllocatedPtr<F>> {
+        Ok(AllocatedPtr::from_parts(
+            &self.get_or_alloc_num(cs, z_ptr.tag.to_field())?,
+            &self.get_or_alloc_num(cs, z_ptr.hash)?,
+        ))
+    }
+}
+
+#[derive(Default)]
+struct SlotData {
+    max_slots: usize,
+    constraints_data: Vec<(Boolean, usize, MetaPtr, Tag)>,
+}
+
+#[derive(Default)]
+struct HashSlots<F: LurkField> {
+    hash2_alloc: Vec<(usize, AllocatedPtr<F>, AllocatedPtr<F>)>,
```

Comment:
(somewhat strong) Opinion: having all these tuples everywhere just obscures the meaning.

Question: Isn't this just the slot allocations?

If so, and if they are sequential as expected, what's the `usize` for? If it turns out that the `usize` element of that tuple is always equal to the tuple's position in the enclosing vector, it's probably not needed. If that's not the case, I think I'm still a little fuzzy on the slot allocation algorithm.